File: scripts/generate-ar-marker.mjs
#!/usr/bin/env node
/** Regenerate ARUCO_4X4_1000 markers for js-aruco2 (bit1=white). */
import fs from 'fs'
import zlib from 'zlib'
import { createRequire } from 'module'
const require = createRequire(import.meta.url)
require('../node_modules/js-aruco2/src/cv.js')
const { AR } = require('../node_modules/js-aruco2/src/aruco.js')
require('../node_modules/js-aruco2/src/dictionaries/aruco_4x4_1000.js')

function crc32(buf) {
  let c = ~0
  for (let i = 0; i < buf.length; i++) {
    c ^= buf[i]
    for (let k = 0; k < 8; k++) c = (c >>> 1) ^ (0xedb88320 & -(c & 1))
  }
  return ~c >>> 0
}
function chunk(type, data) {
  const len = Buffer.alloc(4)
  len.writeUInt32BE(data.length)
  const t = Buffer.from(type)
  const crc = Buffer.alloc(4)
  crc.writeUInt32BE(crc32(Buffer.concat([t, data])))
  return Buffer.concat([len, t, data, crc])
}
function encodePNG(width, height, rgba) {
  const rows = Buffer.alloc((width * 4 + 1) * height)
  for (let y = 0; y < height; y++) {
    rows[y * (width * 4 + 1)] = 0
    rgba.copy(rows, y * (width * 4 + 1) + 1, y * width * 4, (y + 1) * width * 4)
  }
  const ihdr = Buffer.alloc(13)
  ihdr.writeUInt32BE(width, 0)
  ihdr.writeUInt32BE(height, 4)
  ihdr[8] = 8
  ihdr[9] = 6
  return Buffer.concat([
    Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]),
    chunk('IHDR', ihdr),
    chunk('IDAT', zlib.deflateSync(rows, { level: 9 })),
    chunk('IEND', Buffer.alloc(0)),
  ])
}
function bitsFromId(id) {
  const dic = new AR.Dictionary('ARUCO_4X4_1000')
  const code = dic.codeList[id]
  const grid = []
  for (let r = 0; r < 4; r++) {
    const row = []
    for (let c = 0; c < 4; c++) row.push(Number(code[r * 4 + c]))
    grid.push(row)
  }
  return grid
}
function drawArucoRGB(bits, sizePx) {
  const n = 4
  const border = 1
  const cells = n + border * 2
  const cell = Math.floor(sizePx / cells)
  const size = cell * cells
  const rgba = Buffer.alloc(size * size * 4)
  for (let y = 0; y < size; y++)
    for (let x = 0; x < size; x++) {
      const cx = Math.floor(x / cell)
      const cy = Math.floor(y / cell)
      let black
      if (cx === 0 || cy === 0 || cx === cells - 1 || cy === cells - 1) black = true
      else black = bits[cy - border][cx - border] === 0
      const i = (y * size + x) * 4
      const v = black ? 0 : 255
      rgba[i] = rgba[i + 1] = rgba[i + 2] = v
      rgba[i + 3] = 255
    }
  return { size, rgba }
}
function fill(rgba, w, h, color) {
  for (let i = 0; i < w * h; i++) {
    rgba[i * 4] = color[0]
    rgba[i * 4 + 1] = color[1]
    rgba[i * 4 + 2] = color[2]
    rgba[i * 4 + 3] = 255
  }
}
function blit(dst, dw, src, sw, sh, dx, dy) {
  for (let y = 0; y < sh; y++)
    for (let x = 0; x < sw; x++) {
      const si = (y * sw + x) * 4
      const di = ((dy + y) * dw + (dx + x)) * 4
      dst[di] = src[si]
      dst[di + 1] = src[si + 1]
      dst[di + 2] = src[si + 2]
      dst[di + 3] = 255
    }
}

const DPI = 300
const mm = (v) => Math.round((v / 25.4) * DPI)
const PAPER = [245, 247, 248]
const ACCENT = [224, 154, 106]
const OUT = new URL('../public/markers/', import.meta.url)

{
  const bits = bitsFromId(0)
  const { size, rgba } = drawArucoRGB(bits, mm(90))
  fs.writeFileSync(new URL('eyepaint-aruco-id0.png', OUT), encodePNG(size, size, rgba))
}
{
  const W = mm(150)
  const H = mm(150)
  const rgba = Buffer.alloc(W * H * 4)
  fill(rgba, W, H, PAPER)
  const { size, rgba: m } = drawArucoRGB(bitsFromId(0), mm(70))
  const quiet = Math.round(size * 0.12)
  const frame = size + quiet * 2
  const plate = Buffer.alloc(frame * frame * 4)
  fill(plate, frame, frame, [255, 255, 255])
  blit(plate, frame, m, size, size, quiet, quiet)
  const ox = Math.floor((W - frame) / 2)
  const oy = Math.floor((H - frame) / 2) + 30
  blit(rgba, W, plate, frame, frame, ox, oy)
  fs.writeFileSync(new URL('eyepaint-ar-marker-card.png', OUT), encodePNG(W, H, rgba))
}
{
  const W = mm(210)
  const H = mm(297)
  const rgba = Buffer.alloc(W * H * 4)
  fill(rgba, W, H, PAPER)
  const { size, rgba: m } = drawArucoRGB(bitsFromId(0), mm(90))
  const quiet = Math.round(size * 0.18)
  const frame = size + quiet * 2
  const plate = Buffer.alloc(frame * frame * 4)
  fill(plate, frame, frame, [255, 255, 255])
  blit(plate, frame, m, size, size, quiet, quiet)
  const ox = Math.floor((W - frame) / 2)
  const oy = Math.floor(H * 0.28)
  blit(rgba, W, plate, frame, frame, ox, oy)
  const csize = mm(28)
  const margin = mm(18)
  const corners = [
    [margin, margin, 1],
    [W - margin - csize, margin, 2],
    [margin, H - margin - csize, 3],
    [W - margin - csize, H - margin - csize, 4],
  ]
  for (const [x, y, id] of corners) {
    const { size: s, rgba: cm } = drawArucoRGB(bitsFromId(id), csize)
    blit(rgba, W, cm, s, s, x, y)
  }
  fs.writeFileSync(new URL('eyepaint-ar-marker-a4.png', OUT), encodePNG(W, H, rgba))
}
fs.writeFileSync(
  new URL('marker.json', OUT),
  JSON.stringify(
    {
      name: 'EYEPAINT AR marker',
      dictionary: 'ARUCO_4X4_1000',
      detector: 'js-aruco2',
      centerId: 0,
      cornerIds: [1, 2, 3, 4],
      centerSizeMm: 90,
      cornerSizeMm: 28,
      sheet: 'eyepaint-ar-marker-a4.png',
      card: 'eyepaint-ar-marker-card.png',
      solo: 'eyepaint-aruco-id0.png',
      print: 'A4, 100% scale, matte paper preferred',
    },
    null,
    2,
  ) + '\n',
)
console.log('markers regenerated')

File: scripts/generate-ar-marker.py
#!/usr/bin/env python3
"""Generate EYEPAINT printable ArUco marker sheet + card."""

from __future__ import annotations

import json
from pathlib import Path

import cv2
import numpy as np
from PIL import Image, ImageDraw, ImageFont

ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / 'public' / 'markers'

DPI = 300
DICT_ID = cv2.aruco.DICT_4X4_50
CENTER_ID = 0
CORNER_IDS = (1, 2, 3, 4)
INK = '#1C2428'
ACCENT = '#E09A6A'
MUTED = '#5A656B'
PAPER = '#F5F7F8'


def load_font(size: int, bold: bool = False) -> ImageFont.ImageFont:
    names = (
        'DejaVuSans-Bold.ttf' if bold else 'DejaVuSans.ttf',
        'LiberationSans-Bold.ttf' if bold else 'LiberationSans-Regular.ttf',
        'FreeSansBold.ttf' if bold else 'FreeSans.ttf',
    )
    roots = (
        Path('/usr/share/fonts/truetype/dejavu'),
        Path('/usr/share/fonts/truetype/liberation'),
        Path('/usr/share/fonts/truetype/freefont'),
    )
    for root in roots:
        for name in names:
            path = root / name
            if path.exists():
                return ImageFont.truetype(str(path), size)
    return ImageFont.load_default()


def mm(px_per_mm: float, value: float) -> int:
    return int(value * px_per_mm)


def make_aruco(dictionary, marker_id: int, size_px: int) -> Image.Image:
    arr = cv2.aruco.generateImageMarker(dictionary, marker_id, size_px, borderBits=1)
    return Image.fromarray(arr).convert('RGB')


def centered_text(
    draw: ImageDraw.ImageDraw,
    y: float,
    text: str,
    font: ImageFont.ImageFont,
    fill: str,
    width: int,
) -> int:
    bb = draw.textbbox((0, 0), text, font=font)
    tw, th = bb[2] - bb[0], bb[3] - bb[1]
    draw.text(((width - tw) / 2, y), text, fill=fill, font=font)
    return th


def build_a4(dictionary) -> Image.Image:
    w = mm(DPI / 25.4, 210)
    h = mm(DPI / 25.4, 297)
    canvas = Image.new('RGB', (w, h), PAPER)
    draw = ImageDraw.Draw(canvas)

    font_brand = load_font(140, bold=True)
    font_sub = load_font(42)
    font_hint = load_font(36)
    font_meta = load_font(28)
    font_small = load_font(24)

    brand_y = int(h * 0.08)
    bh = centered_text(draw, brand_y, 'EYEPAINT', font_brand, INK, w)
    uy = brand_y + bh + 28
    draw.rectangle([w * 0.35, uy, w * 0.65, uy + 10], fill=ACCENT)
    centered_text(draw, uy + 36, 'Маркер плоскости AR', font_sub, MUTED, w)

    marker_mm = 90
    marker_px = mm(DPI / 25.4, marker_mm)
    quiet = int(marker_px * 0.22)
    plate = Image.new('RGB', (marker_px + quiet * 2, marker_px + quiet * 2), '#FFFFFF')
    plate.paste(make_aruco(dictionary, CENTER_ID, marker_px), (quiet, quiet))
    outer = Image.new('RGB', (plate.width + 24, plate.height + 24), '#FFFFFF')
    outer.paste(plate, (12, 12))
    ImageDraw.Draw(outer).rectangle(
        [0, 0, outer.width - 1, outer.height - 1],
        outline=ACCENT,
        width=6,
    )

    mx = (w - outer.width) // 2
    my = int(h * 0.28)
    canvas.paste(outer, (mx, my))
    centered_text(
        draw,
        my + outer.height + 36,
        f'ArUco 4×4  ·  ID {CENTER_ID}  ·  {marker_mm} мм',
        font_meta,
        MUTED,
        w,
    )

    hints = [
        '1. Распечатай лист без масштаба (100% / Actual size)',
        '2. Положи маркер на стол / лист бумаги',
        '3. В EYEPAINT: режим AR → дождись фиксации',
        '4. Убери маркер в сторону — референс останется на плоскости',
        '5. Камера сдвинулась → «Перекалибровать»',
    ]
    hy = my + outer.height + 120
    for i, line in enumerate(hints):
        centered_text(draw, hy + i * 58, line, font_hint, INK, w)

    corner_px = mm(DPI / 25.4, 28)
    margin = mm(DPI / 25.4, 18)
    corners = [
        (margin, margin, CORNER_IDS[0]),
        (w - margin - corner_px, margin, CORNER_IDS[1]),
        (margin, h - margin - corner_px, CORNER_IDS[2]),
        (w - margin - corner_px, h - margin - corner_px, CORNER_IDS[3]),
    ]
    for x, y, mid in corners:
        canvas.paste(make_aruco(dictionary, mid, corner_px), (x, y))

    centered_text(
        draw,
        h - margin - corner_px - 70,
        'maplol/Eyepaint  ·  dict DICT_4X4_50  ·  center ID 0  ·  corners 1–4',
        font_small,
        MUTED,
        w,
    )
    centered_text(
        draw,
        h - margin - corner_px - 36,
        'Формат A4 · печать односторонняя · не ламинировать глянцем (блики мешают)',
        font_small,
        MUTED,
        w,
    )
    return canvas


def build_card(dictionary) -> Image.Image:
    side = mm(DPI / 25.4, 150)
    card = Image.new('RGB', (side, side), PAPER)
    draw = ImageDraw.Draw(card)
    font_brand = load_font(72, bold=True)
    font_sub = load_font(28)

    bh = centered_text(draw, int(side * 0.08), 'EYEPAINT', font_brand, INK, side)
    draw.rectangle(
        [side * 0.32, int(side * 0.08) + bh + 16, side * 0.68, int(side * 0.08) + bh + 24],
        fill=ACCENT,
    )

    marker_mm = 70
    marker_px = mm(DPI / 25.4, marker_mm)
    quiet = int(marker_px * 0.2)
    plate = Image.new('RGB', (marker_px + quiet * 2, marker_px + quiet * 2), '#FFFFFF')
    plate.paste(make_aruco(dictionary, CENTER_ID, marker_px), (quiet, quiet))
    cx = (side - plate.width) // 2
    cy = (side - plate.height) // 2 + 20
    card.paste(plate, (cx, cy))
    centered_text(
        draw,
        cy + plate.height + 28,
        'AR · ID 0 · убери после фиксации',
        font_sub,
        MUTED,
        side,
    )
    return card


def main() -> None:
    OUT.mkdir(parents=True, exist_ok=True)
    dictionary = cv2.aruco.getPredefinedDictionary(DICT_ID)

    a4 = build_a4(dictionary)
    a4_path = OUT / 'eyepaint-ar-marker-a4.png'
    a4.save(a4_path, 'PNG', dpi=(DPI, DPI))

    card = build_card(dictionary)
    card_path = OUT / 'eyepaint-ar-marker-card.png'
    card.save(card_path, 'PNG', dpi=(DPI, DPI))

    solo_px = mm(DPI / 25.4, 90)
    solo = make_aruco(dictionary, CENTER_ID, solo_px).convert('L')
    solo_path = OUT / 'eyepaint-aruco-id0.png'
    solo.save(solo_path, 'PNG')

    meta = {
        'name': 'EYEPAINT AR marker',
        'dictionary': 'DICT_4X4_50',
        'opencvDictId': int(DICT_ID),
        'centerId': CENTER_ID,
        'cornerIds': list(CORNER_IDS),
        'centerSizeMm': 90,
        'cornerSizeMm': 28,
        'sheet': a4_path.name,
        'card': card_path.name,
        'solo': solo_path.name,
        'print': 'A4, 100% scale, matte paper preferred',
    }
    (OUT / 'marker.json').write_text(json.dumps(meta, ensure_ascii=False, indent=2) + '\n')
    print('wrote', a4_path)
    print('wrote', card_path)
    print('wrote', solo_path)


if __name__ == '__main__':
    main()
